File: packages/shared/src/arrhenius/arrhenius.ts
import {
  ACTIVATION_ENERGY_EV,
  BOLTZMANN_EV_PER_K,
  DEFAULT_DIFFUSION_COEFF_CM2_PER_S,
} from "../constants/physical";
import type { ArrheniusInput } from "../types/simulation";

export function calculateDiffusionCoeffArrhenius(input: ArrheniusInput): number {
  const { diffusionRefCm2PerS, temperatureRefK, temperatureK, activationEnergyEv } = input;
  if (temperatureK <= 0 || temperatureRefK <= 0) {
    throw new Error("Temperature must be positive in Kelvin");
  }
  const exponent =
    (-activationEnergyEv / BOLTZMANN_EV_PER_K) * (1 / temperatureK - 1 / temperatureRefK);
  return diffusionRefCm2PerS * Math.exp(exponent);
}

export const DEFAULT_ARRHENIUS_INPUT: ArrheniusInput = {
  diffusionRefCm2PerS: DEFAULT_DIFFUSION_COEFF_CM2_PER_S,
  temperatureRefK: 298.15,
  temperatureK: 298.15,
  activationEnergyEv: ACTIVATION_ENERGY_EV,
};

File: packages/diffusion-core/src/arrhenius.rs
use crate::domain::ArrheniusInput;

const BOLTZMANN_EV_PER_K: f64 = 8.617_333_262e-5;

pub fn calculate_diffusion_coeff_arrhenius(input: &ArrheniusInput) -> Result<f64, String> {
    if input.temperature_k <= 0.0 || input.temperature_ref_k <= 0.0 {
        return Err("temperature must be positive in Kelvin".to_string());
    }
    let exponent = (-input.activation_energy_ev / BOLTZMANN_EV_PER_K)
        * (1.0 / input.temperature_k - 1.0 / input.temperature_ref_k);
    Ok(input.diffusion_ref_cm2_per_s * exponent.exp())
}
